<a href="https://colab.research.google.com/github/KCM2005/Kalenga-Mumba/blob/main/HW_14_TextGeneration.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Final homework assignment:  generating text from a corpus

Due Friday 3 May at 5 pm

We begin by loading all of Shakespeare's plays into a text file, and separating them into individual words.  This is made easy by the fact that someone already made a giant text file of these plays for us, and put it online.

[Note that by changing the url in the following code chunk, you can load any text file that lives online in this way . . . see below.]

Run the following code.  It should output a list of ten words, starting with 'From' and ending with 'rose'.

In [1]:
from urllib.request import urlopen            #package for reading info from the web
url='https://ocw.mit.edu/ans7870/6/6.006/s08/lecturenotes/files/t8.shakespeare.txt'   #web address
text = urlopen(url).read().decode('utf-8')    #open web address, read file, convert to text
shakespeare_words=text.split()[1651:]         #split text into words (not including preamble)
shakespeare_words[:10]                        #display the first 10 words

['From',
 'fairest',
 'creatures',
 'we',
 'desire',
 'increase,',
 'That',
 'thereby',
 "beauty's",
 'rose']

## 1. `nextwords` practice

Create a practice function called `createNextwordsPractice` which takes a list of words called `wordlist` and returns a dictionary called `nextwordsPractice`. The keys of `nextwordsPractice` should be all the words in `wordlist`, and the value for any given key should be a list of all the words that follow the key word somewhere in the original `wordlist`.



In [100]:
def createNextwordsPractice(wordlist):
  nextwordsPractice = {}

  for i in range(len(wordlist) - 1):
    currentword = wordlist[i]
    nextword = wordlist[i + 1]

    if currentword not in nextwordsPractice:
      nextwordsPractice[currentword] = nextword
    else:
      nextwordsPractice[currentword].append(nextword)

  return nextwordsPractice

wordlist = ['forest', 'wind', 'sun', 'heart', 'jealous', 'interest']
createNextwordsPractice(wordlist)

{'forest': 'wind',
 'wind': 'sun',
 'sun': 'heart',
 'heart': 'jealous',
 'jealous': 'interest'}

Run the following code:

In [3]:
practice = createNextwordsPractice(shakespeare_words)
practice['From']

['fairest',
 'his',
 'sullen',
 'limits',
 'hands',
 'whence',
 'where',
 'me',
 'this',
 'thy',
 'hence',
 'thee,',
 'you',
 'heaven',
 'her',
 'simple',
 'whence',
 'below',
 'lowest',
 'courtly',
 'son',
 'the',
 'the',
 'Sicyon,',
 'Alexandria',
 'Silvius,',
 'mine',
 'the',
 'firm',
 'Egypt',
 'Antony',
 'thine',
 'that',
 'my',
 'me',
 'henceforth',
 'tyrant',
 'seventeen',
 'miserable',
 'whom',
 'my',
 'whence,',
 'me',
 'me',
 'slaves',
 'every',
 'rascals',
 'whom',
 'where',
 "th'",
 'Rome',
 'off',
 'these',
 'twelve',
 "th'",
 'whence?',
 'Rome.',
 'thy',
 'Gallia',
 'fairies',
 'off',
 "th'",
 'most',
 'thy',
 'youth',
 'every',
 'whose',
 'so',
 'one',
 'this',
 'the',
 'east',
 'action',
 'my',
 'this',
 'Leonati',
 'her',
 'chance',
 'south',
 'our',
 'whence',
 'the',
 'top',
 'this',
 'our',
 'that',
 'me,',
 'fashion',
 'the',
 'another',
 'Hamlet?',
 'whence',
 'whom',
 "Rumour's",
 'whence',
 'the',
 'a',
 'a',
 'enemies',
 'every',
 'this',
 'the',
 'open',
 'who

In [4]:
practice['rose']

['might',
 'looks',
 'is',
 'of',
 'Of',
 'may',
 'will',
 'of',
 'From',
 'and',
 'of',
 'both',
 'with',
 'from',
 'with',
 'with',
 'side.',
 'red,',
 'too.',
 'in',
 'a',
 'a',
 'I',
 'against',
 'that',
 'and',
 'from',
 'again,',
 'Lest',
 'against',
 'Than',
 "distill'd",
 'on',
 'up',
 'in',
 'wither.',
 'and',
 'By',
 'again?',
 'he']

In [5]:
practice['rose.']

['It']

## 2. `nextwords` for real:  new sentences.

Next, write a new function called `createNextwords` which takes a `wordlist` and returns a dictionary of `nextwords`, but now with a new additional feature.  
In addition to the keys from problem 1, add a special key to the dictionary, which will be the symbol '%'.  This key will signify the first word of a sentence. When scanning the wordlist to construct the `nextwords` dictionary, normally you would add `wordlist[i+1]` to the list `nextwords[wordlist[i]]`. However, if `wordlist[i]` ends with a period, then you should instead add `wordlist[i+1]` to the list nextwords['%']. Also add the very first word of `wordlist` to nextwords['%'].


You should be able to produce this output:
```
> shakespeare_nextwords=createNextwords(shakespeare_words)
> shakespeare_nextwords['Rose']
['at', 'or', 'or', 'and']

> shakespeare_nextwords['rose.']
[]

> len(shakespeare_nextwords['%'])
76543
```

In [95]:
def createNextwords(wordlist):
  nextwords = {'%': []}
  for i in range(len(wordlist)):
    currentword = wordlist[i]

    if i == 0:
      nextword = '%'
    else:
      nextword = wordlist[i - 1]

    if wordlist[i].endswith('.'):
      nextwords['%'].append(wordlist[i + 1])
    else:
      if currentword not in nextwords:
        nextwords[currentword] = [nextword]
      else:
        nextwords[currentword].append(nextword)
  return nextwords

wordlist = ['absolute', 'front', 'honest', 'simular']
createNextwords(wordlist)

{'%': [],
 'absolute': ['%'],
 'front': ['absolute'],
 'honest': ['front'],
 'simular': ['honest']}

Run the following code:

In [7]:
shakespeare_nextwords = createNextwords(shakespeare_words)

print(shakespeare_nextwords['Rose'])
print(len(shakespeare_nextwords['%']))

['together,', 'White', 'Red', 'Red']
76542


In [8]:
print(len(shakespeare_nextwords['rose']))

40


In [9]:
print(shakespeare_nextwords['rose'])

["beauty's", 'The', 'his', 'our', 'the', 'blown', 'sweetest', 'and', 'the', 'he', 'O', 'we', 'white', 'red', 'white', 'red', 'white', 'white', 'white', 'white', 'thy', 'thy', 'the', 'He', 'white', 'red', 'red', 'Then', 'a', 'Brutus', 'a', 'the', 'red', 'they', 'a', 'fair', 'white', 'a', 'he', 'sun']


## 3.  Make your own Shakespeare play

Okay, that may be a bit of an exaggeration.  Anyway, next up is text generation.  
We'll need the choice function:

In [15]:
from random import choice
import random

This function chooses a random element from a list.  Run this code a few times:

In [16]:
choice([1, 17, 34, 289])

1

Now, create a function called `makePlay(nextwords,n)` which takes a dictionary `nextwords` and a number `n`, and follows these steps:
1. Define an empty `newplay` string.  (I did this for you.)
2. Set `currentword` equal to '%'.
3. Start a for-loop that will run `n` times.
4. Randomly choose a word from the list `nextwords[currentword]`.
5. Add the chosen word to `newplay`, followed by a space `' '` character.
6. If the chosen word ends in a period, add a newline character '\n' and set `currentword` equal to `%`.
7. Otherwise, set `currentword` equal to the chosen word.
8. Go back to step 4.
9. When all `n` iterations have finished, your function should return `newplay`.

In [29]:
def makePlay(nextwords,n):
  newplay=''
  currentword = '%'

  for i in range(n):
    next_words = nextwords[currentword]
    chosen_word = random.choice(next_words)
    newplay += chosen_word + ' '

    if chosen_word.endswith('.'):
      newplay += '\n'
      currentword = '%'
    else:
      currentword = chosen_word
  return newplay


Now create a 1000 word Shakesperian play by running the following code:

In [58]:
url='https://folger-main-site-assets.s3.amazonaws.com/uploads/2022/11/romeo-and-juliet_TXT_FolgerShakespeare.txt'
text = urlopen(url).read().decode('utf-8')
wordlist = text.split()[24506:5]
wordlist[:20]
nextwords=createNextwords(shakespeare_words)
newplay=makePlay(nextwords,1000)
print(newplay)

Earth doth so, you to bed. Faith, you'll be brief. O, by the Prince's doom. ROMEO This gentleman, and purge Myself condemned and another lends them women of packthread, and trim When and Romeo.] Within this shall forget, to Juliet and another man, And soar with another's burning; One who, to that there for shame! Confusion's cure lives married with dead before, Warm and love in this uttered With twenty could find thee that thou but what a tithe-pig's tail, Tickling a well, nor proud you shall no Cupid wings. Now do bear thee in our morning with sorrow will never die, Take heed, for a wish! He lent us all those two days doth not a poor and logs and when I say well. FRIAR LAWRENCE, [aside] For hark you, Thank me not, Your looks bid "Good morrow" to fire; And then dreams of that dost not trouble Peter for some mean these walls, For doting, not rosemary On Thursday, tell them merrily. If thou our drift, And shake the slip. Can you Here lies And usest none in triumph, and nothing can read. 

At this point you might enjoy finding other sources of text on the internet and replacing the url at the top of this notebook. You can use song lyrics, news articles, etc. You just need to find files ending with '.txt'. Alternatively, you can copy and paste any block of text from any file into colab, assign it to the variable `text`, and then use wordlist=text.split() to get a list of words.

## Bonus:

Recall the following recursive function that we used to define the Fibonacci numbers:

In [99]:
def fibonacci(n):
  # Base cases: F(0) = 0, F(1) = 1
  if n == 0:
    return 0
  elif n == 1:
    return 1
  else:
    # Recursive case: F(n) = F(n-1) + F(n-2)
    return fibonacci(n - 1) + fibonacci(n - 2)

fibonacci(17)

#I think the fibonacci() function gets called about 20 times

1597

Here is the question.  If you run this line of code:  
`fibonacci(17)`  
how many times does the function `fibonacci()` get called?  
(The answer is not 1597, which is the value returned by the function.)